pymc3/gp/gp.py

```diff
@@ -843,98 +844,104 @@ def __init__(self, mean_func=Zero(), cov_funcs=[Constant(0.0)]):
-        #     noise = pm.gp.cov.WhiteNoise(noise)
+        self._check_inputs(Xs, y, sigma)
         mu, covs = self._build_marginal_likelihood(Xs)
         self.Xs = Xs
         self.y = y
-        self.noise = noise
+        self.sigma = sigma
         if is_observed:
-            return pm.KroneckerNormal(name, mu=mu, covs=covs, noise=noise,
+            return pm.KroneckerNormal(name, mu=mu, covs=covs, noise=sigma,
                                       observed=y, **kwargs)
         else:
-            # shape = infer_shape(Xs, kwargs.pop("shape", None))
-            return pm.KroneckerNormal(name, mu=mu, covs=covs, noise=noise,
-                                      **kwargs)
+            shape = np.prod([len(X) for X in Xs])
+            return pm.KroneckerNormal(name, mu=mu, covs=covs, noise=sigma,
+                                      shape=shape, **kwargs)
 
     def _get_given_vals(self, given):
         if given is None:
             given = {}
 
         if 'gp' in given:
             cov_total = given['gp'].cov_func
+            cov_totals = given['gp'].cov_funcs
             mean_total = given['gp'].mean_func
         else:
             cov_total = self.cov_func
+            cov_totals = self.cov_funcs
             mean_total = self.mean_func
-        if all(val in given for val in ['X', 'y', 'noise']):
-            X, y, noise = given['X'], given['y'], given['noise']
-            # if not isinstance(noise, Covariance):
-            #     noise = pm.gp.cov.WhiteNoise(noise)
+        if all(val in given for val in ['Xs', 'y', 'sigma']):
+            Xs, y, sigma = given['Xs'], given['y'], given['sigma']
         else:
-            X, y, noise = self.X, self.y, self.noise
-        return X, y, noise, cov_total, mean_total
+            Xs, y, sigma = self.Xs, self.y, self.sigma
+        self._check_inputs(Xs, y, sigma)
+        return Xs, y, sigma, cov_total, cov_totals, mean_total
 
-    def _build_conditional(self, Xnew, pred_noise, diag, Xs, y, noise,
-                           cov_total, mean_total):
+    def _build_conditional(self, Xnew, pred_noise, diag, Xs, y, sigma,
+                           cov_total, cov_totals, mean_total):
         # Old points
-        delta = y - self.mean_func(cartesian(*Xs))
-        Kns = [f(X) for f, X in zip_longest(self.cov_funcs, Xs)]
-        eigs_sep, Qs = zip(*map(np.linalg.eigh, Kns))  # Unzip
-        QTs = list(map(np.transpose, Qs))
+        X = cartesian(*Xs)
+        delta = y - mean_total(X)
+        Kns = [f(x) for f, x in zip(cov_totals, Xs)]
+        eigs_sep, Qs = zip(*map(eigh, Kns))  # Unzip
+        QTs = list(map(tt.transpose, Qs))
         eigs = kron_diag(eigs_sep)  # Combine separate eigs
-        if noise is not None:
-            eigs += noise
+        if sigma is not None:
+            eigs += sigma
 
         # New points
-        Km = self.total_cov_func(Xnew, diag)
-        Knm = self.total_cov_func(cartesian(*Xs), Xnew)
+        Km = cov_total(Xnew, diag)
```

Comment:
`cov_total` is here to support making additive models, so that a component GP can be used in isolation to make predictions (described [here](http://docs.pymc.io/gp.html#additive-gps)).  It may not be straightforward to support this for Kronecker GPs, in which case it would be good to implement an `__add__` method that raises an error, like in the `TP` class.

Reply:
You're right, the sum of Kronecker-structured covariance matrices does not in general have a Kronecker structure, so this class wouldn't be closed under addition. In fact, the `conditional` method doesn't yield a `KroneckerNormal` either, so technically that could be inherited from `Marginal` if we didn't care about the speedups from `_build_conditional`.

Reply:
Good call.  I've seen this structure [once in a paper](https://papers.nips.cc/paper/5089-it-is-all-in-the-noise-efficient-multi-task-gaussian-process-inference-with-structured-residuals), but definitely out of the scope of this PR

Reply:
That's an interesting paper! It's too bad that the speedups only hold for the sum of two terms.